streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicConfig.java

```diff
@@ -67,33 +82,43 @@ private boolean isCompactDelete() {
      * @param additionalRetentionMs - added to retention to allow for clock drift etc
      * @return Properties to be used when creating the topic
      */
-    public Properties toProperties(final long additionalRetentionMs) {
-        final Properties result = new Properties();
-        for (Map.Entry<String, String> configEntry : logConfig.entrySet()) {
-            result.put(configEntry.getKey(), configEntry.getValue());
-        }
-        if (retentionMs != null && isCompactDelete()) {
-            result.put(InternalTopicManager.RETENTION_MS, String.valueOf(retentionMs + additionalRetentionMs));
+    public Map<String, String> getProperties(final Map<String, String> defaultProperties, final long additionalRetentionMs) {
+        // internal topic config overridden rule: library overrides < global config overrides < per-topic config overrides
+        final Map<String, String> topicConfig = new HashMap<>();
+
+        switch (topicType) {
```

Comment:
I think i said this before, but i think it would be better/nicer to have different implementations of InternalTopicConfig for the different types of internal topics and then we don't need to have `switch` and `if` statements. Each impl will know what ti needs to do. WDYT?

Reply:
I thought about it but felt it may be an overkill.. because now we have all the logics depending on `type` consolidated in one class already.

Reply:
I personally support Damian idea here. Don't think it's overkill but actually more elegant OO code. If you feel strong to keep as it, also fine with me though.

Reply:
@dguy @mjsax  Just adding a few more notes here: I think the major concern is our type-dependent logic is scattered across multiple packages and classes, so we'd like to consolidate it. One way is to still pass in the type information but make all related logic within this single class (what I did), another way is to make different instantiations of the class so we do not need to have this `type` field at all. Personally I think they achieved similar goal while mine is less code: the downside is, as mentioned in https://github.com/apache/kafka/pull/4315#discussion_r156749428 the `setRetentionMs` should only be accepted with `windowed store changelog` type, but we still need to provide this API for either approach.

Reply:
> I personally support Damian idea here. Don't think it's overkill but actually more elegant OO code. If you feel strong to keep as it, also fine with me though.

I did not see this comment while writing mine. If my understanding about the root concern is right but we feel it is better to get rid of this `type` information, I can change it (i.e. I do not feel strong about it :P). LMK @mjsax @dguy

Reply:
I don't think it is overkill either, it is just good OO design. Anyway, i'm not going to block this PR because of it.

Reply:
Cool, I have refactored `InternalTopicConfig` in the latest commit as you suggested and will go ahead to merge it as is; if you can take a quick look that would be greatly appreciated.